cpp/src/rolling/detail/range_utils.cuh

```diff
@@ -0,0 +1,869 @@
+/*
+ * Copyright (c) 2024-2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "rolling_utils.cuh"
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/rolling.hpp>
+#include <cudf/rolling.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_checks.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+#include <rmm/resource_ref.hpp>
+
+#include <cuda/functional>
+#include <cuda/std/limits>
+#include <cuda/std/type_traits>
+#include <thrust/binary_search.h>
+#include <thrust/copy.h>
+#include <thrust/distance.h>
+#include <thrust/execution_policy.h>
+
+#include <optional>
+#include <variant>
+
+namespace CUDF_EXPORT cudf {
+namespace detail {
+namespace rolling {
+
+/**
+ * @brief Indicates behaviour of boundary of a rolling window.
+ */
+enum class window_tag : std::int8_t {
+  BOUNDED_OPEN,    ///< Window is bounded by a value-based endpoint, endpoint is excluded.
+  BOUNDED_CLOSED,  ///< Window is bounded by a value-based endpoint, endpoint is included.
+  UNBOUNDED,       ///< Window runs to beginning (or end) of the group the row is in.
+  CURRENT_ROW,     ///< Window contains all rows that compare equal to the current row.
```

Comment:
I suppose I can actually use the public types and do constexpr dispatch via `is_same_v` rather than comparison to a compile-time known enum tag.